### PR TITLE
Log success rate for on policy algorithms

### LINKED ISSUE #1867: [Bug]: `rollout/success_rate` does not show for Monitor + OnPolicyAlgorithm
### 🐛 Bug

# Expected Behavior
Based on [SB3 Documentation](https://stable-baselines3.readthedocs.io/en/master/common/logger.html#rollout), `success_rate` should display for a `Monitor-wrapped Environment` + `OnPolicyAlgorithm` if:
> "...an extra argument to the Monitor wrapper [is passed] to log that value `(info_keywords=("is_success",))` and `info["is_success"]=True/False` on the final step of the episode [is provided]."

# Actual Behavior

According to actual runs of the environment (_see log output_) and [source code](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/on_policy_algorithm.py#L253), `success_rate` is not actually recorded.

# Extra Info
`OffPolicyAlgorithm` [appears to behave as expected](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/off_policy_algorithm.py#L429).

### Code example

```python
from stable_baselines3.common.monitor import Monitor
from stable_baselines3.common.env_checker import check_env
from stable_baselines3.ppo.ppo import PPO

from gymnasium.envs.classic_control.cartpole import CartPoleEnv

class CartPoleWithSuccess(CartPoleEnv):
    def step(self, action):
        state, reward, terminated, truncated, info = super().step(action)
        if terminated or truncated: 
            info["is_success"] = (-self.x_threshold < self.state[0] < self.x_threshold
                                    and -self.theta_threshold_radians < self.state[2] < self.theta_threshold_radians
                                    )
        return state, reward, terminated, truncated, info

if __name__ == "__main__":
    env = Monitor(CartPoleWithSuccess(), filename = "tmp/log.csv", info_keywords = ("is_success", ))
    check_env(env)

    model = PPO("MlpPolicy", env = env, stats_window_size = 10, verbose = 1)
    model.learn(total_timesteps = int(5e5), progress_bar = True)
```

### Relevant log output / Error message

```shell
Using cuda device
Wrapping the env in a DummyVecEnv.
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 21.1     |
|    ep_rew_mean     | 21.1     |
| time/              |          |
|    fps             | 1170     |
|    iterations      | 1        |
|    time_elapsed    | 1        |
|    total_timesteps | 2048     |
---------------------------------
----------------------------------------
| rollout/                |            |
|    ep_len_mean          | 28.5       |
|    ep_rew_mean          | 28.5       |
| time/                   |            |
|    fps                  | 1135       |
|    iterations           | 2          |
|    time_elapsed         | 3          |
|    total_timesteps      | 4096       |
| train/                  |            |
|    approx_kl            | 0.00899566 |
|    clip_fraction        | 0.0829     |
|    clip_range           | 0.2        |
|    entropy_loss         | -0.686     |
|    explained_variance   | 0.00679    |
|    learning_rate        | 0.0003     |
|    loss                 | 6.37       |
|    n_updates            | 10         |
|    policy_gradient_loss | -0.0119    |
|    value_loss           | 44.4       |
----------------------------------------
-----------------------------------------
| rollout/                |             |
|    ep_len_mean          | 50          |
|    ep_rew_mean          | 50          |
| time/                   |             |
|    fps                  | 1116        |
|    iterations           | 3           |
|    time_elapsed         | 5           |
|    total_timesteps      | 6144        |
| train/                  |             |
|    approx_kl            | 0.008980946 |
|    clip_fraction        | 0.0452      |
|    clip_range           | 0.2         |
|    entropy_loss         | -0.672      |
|    explained_variance   | 0.111       |
|    learning_rate        | 0.0003      |
|    loss                 | 14.7        |
|    n_updates            | 20          |
|    policy_gradient_loss | -0.0137     |
|    value_loss           | 39.4        |
-----------------------------------------
```

### System Info

# Describe the characteristic of your environment:

  -  Describe how the library was installed (pip, docker, source, ...)
       - `pip install stable-baselines3`
  - GPU models and configuration
        - NVIDIA A2000 8GB
  - Python version
        - `3.10.12`
  - PyTorch version
        - `2.0.1+cu117`
  - Gymnasium version
        - `0.29.1`

## Results from `python -c import stable_baselines3 as sb3; sb3.get_system_info()`:
```
- OS: Linux-6.5.0-25-generic-x86_64-with-glibc2.35 # 25~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Feb 20 16:09:15 UTC 2
- Python: 3.10.12
- Stable-Baselines3: 2.2.1
- PyTorch: 2.0.1+cu117
- GPU Enabled: True
- Numpy: 1.25.0
- Cloudpickle: 2.2.1
- Gymnasium: 0.29.1

({'OS': 'Linux-6.5.0-25-generic-x86_64-with-glibc2.35 # 25~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Feb 20 16:09:15 UTC 2', 
'Python': '3.10.12', 
'Stable-Baselines3': '2.2.1', 
'PyTorch': '2.0.1+cu117', 
'GPU Enabled': 'True',
'Numpy': '1.25.0',
'Cloudpickle': '2.2.1', 
'Gymnasium': '0.29.1'}, 
'- OS: Linux-6.5.0-25-generic-x86_64-with-glibc2.35 # 25~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Feb 20 16:09:15 UTC 2\n- Python: 3.10.12\n- Stable-Baselines3: 2.2.1\n- PyTorch: 2.0.1+cu117\n- GPU Enabled: True\n- Numpy: 1.25.0\n- Cloudpickle: 2.2.1\n- Gymnasium: 0.29.1\n')
```

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] I have provided a [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014) example to reproduce the bug
- [X] I have checked my env using the env checker
- [X] I've used the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

### PR DESCRIPTION
**Pull request by corentinlger**
Hi, I changed the ```on_policy_algorithm.py``` file to enable showing rollout/success_rate on the monitor for ```OnPolicyAlgorithm```

## Description

I added ```dones``` as argument to ```self._update_info_buffer``` to effectively update the buffer (before it couldn't save ```info['is_success']``` because dones was set to ```None```.

Then I added these lines to the ones that display training infos on the monitor (as in [off_policy_algorithm.py](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/off_policy_algorithm.py) : 

```python
if len(self.ep_success_buffer) > 0:
            self.logger.record("rollout/success_rate", safe_mean(self.ep_success_buffer))
```

I also refactored the code to put this whole block writing logs in a ```_dump_logs``` function (also in the spirit of [off_policy_algorithm.py](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/off_policy_algorithm.py))

Is the PR ok for you ? If yes I will implement the associated test.

## Motivation and Context

closes #1867
- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [ ] I have opened an associated PR on the [SB3-Contrib repository](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib) (if necessary)
- [ ] I have opened an associated PR on the [RL-Zoo3 repository](https://github.com/DLR-RM/rl-baselines3-zoo) (if necessary)
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: stable_baselines3/common/on_policy_algorithm.py [hunk 1]
```diff
@@ -92,6 +92,7 @@ def __init__(
             use_sde=use_sde,
             sde_sample_freq=sde_sample_freq,
             support_multi_env=True,
+            monitor_wrapper=monitor_wrapper,
```
Comment (araffin) [inline stable_baselines3/common/on_policy_algorithm.py:95]: nice catch =) (we need to mention that in the changelog)

**Review by araffin — COMMENTED** [review]
Thanks for the PR, overall LGTM, please go ahead with the tests =)
(and we would need to have a PR ready for SB3 contrib probably to update MaskablePPO and other PPO derivates if needed)

File: stable_baselines3/common/on_policy_algorithm.py (hunk 1 above)
Comment (corentinlger) [reply stable_baselines3/common/on_policy_algorithm.py:95]: How should we mention it ?

Comment (araffin) [reply stable_baselines3/common/on_policy_algorithm.py:95]: something like "Fixed monitor_wrapper argument that was not passed to the parent class`

**Comment by corentinlger** [comment]
Hi, I added a test with a dummy environment that returns ```True``` or ```False``` for the success info according to a list of dummy successes (i.e a list of size (n_logs, n_ep_per_log) that determines if the ```episode j``` of ```logging iteration i``` is going to be a success or not). I put it in ```tests/test_logger.py```, it seems to work but I don't know if it is what you expected. 

There is just one thing I'm not sure to understand : instead of getting 0.3, 0.5 and 0.8 success_rate on the three logging iterations (according to the dummy success list I manually created), I get 0.3333333333333333, 0.5555555555555556 and 0.7777777777777778. It don't know if I did something wrong or if it is linked to how the success_rate is computed. Do you know why this happens @araffin ?

Also there seems to be an error with the ```check-codestyle``` of ```stable_baselines3/common/vec_env/vec_frame_stack.py``` in the CI pipeline but I didn't touch this file.

**Comment by araffin** [comment]
> It don't know if I did something wrong or if it is linked to how the success_rate is computed. Do you know why this happens @araffin ?

I think this was an off-by-one error in the tests, I fixed it and cleanup the test in https://github.com/DLR-RM/stable-baselines3/pull/1870/commits/241595275124b420d95c716ad720b68ec55bb4d1

Could you do a similar PR for SB3 contrib?

**Review by araffin — APPROVED** [review]
LGTM, thanks =)

**Comment by corentinlger** [comment]
> I think this was an off-by-one error in the tests, I fixed it and cleanup the test in [2415952](https://github.com/DLR-RM/stable-baselines3/commit/241595275124b420d95c716ad720b68ec55bb4d1)

Ok cool ! Indeed the test looks cleaner now (:

> Could you do a similar PR for SB3 contrib?

Sure, I'll try that. I only need to implement and test the feature for ```MaskablePPO``` and ``````RecurrentPPO`````` right ?

**Comment by araffin** [comment]
> Sure, I'll try that. I only need to implement and test the feature for MaskablePPO and RecurrentPPO right ?

yes, although it might not be needed as they derive from the on policy base class.

**Pull request merged**